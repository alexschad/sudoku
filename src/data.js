export const rows = [
  [0, 1, 2, 3, 4, 5, 6, 7, 8],
  [9, 10, 11, 12, 13, 14, 15, 16, 17],
  [18, 19, 20, 21, 22, 23, 24, 25, 26],
  [27, 28, 29, 30, 31, 32, 33, 34, 35],
  [36, 37, 38, 39, 40, 41, 42, 43, 44],
  [45, 46, 47, 48, 49, 50, 51, 52, 53],
  [54, 55, 56, 57, 58, 59, 60, 61, 62],
  [63, 64, 65, 66, 67, 68, 69, 70, 71],
  [72, 73, 74, 75, 76, 77, 78, 79, 80],
];
export const cols = [
  [0, 9, 18, 27, 36, 45, 54, 63, 72],
  [1, 10, 19, 28, 37, 46, 55, 64, 73],
  [2, 11, 20, 29, 38, 47, 56, 65, 74],
  [3, 12, 21, 30, 39, 48, 57, 66, 75],
  [4, 13, 22, 31, 40, 49, 58, 67, 76],
  [5, 14, 23, 32, 41, 50, 59, 68, 77],
  [6, 15, 24, 33, 42, 51, 60, 69, 78],
  [7, 16, 25, 34, 43, 52, 61, 70, 79],
  [8, 17, 26, 35, 44, 53, 62, 71, 80],
];
export const squares = [
  [0, 1, 2, 9, 10, 11, 18, 19, 20],
  [3, 4, 5, 12, 13, 14, 21, 22, 23],
  [6, 7, 8, 15, 16, 17, 24, 25, 26],
  [27, 28, 29, 36, 37, 38, 45, 46, 47],
  [30, 31, 32, 39, 40, 41, 48, 49, 50],
  [33, 34, 35, 42, 43, 44, 51, 52, 53],
  [54, 55, 56, 63, 64, 65, 72, 73, 74],
  [57, 58, 59, 66, 67, 68, 75, 76, 77],
  [60, 61, 62, 69, 70, 71, 78, 79, 80],
];

// prettier-ignore
export const sudokus = [
  [null,null,2,null,null,4,8,9,null,null,5,null,8,null,null,3,null,7,7,null,null,null,5,null,null,4,6,null,6,null,null,null,null,null,null,4,null,null,1,null,9,null,5,null,null,9,null,null,null,null,null,null,8,null,6,3,null,null,1,null,null,null,8,1,null,4,null,null,8,null,2,null,null,9,8,4,null,null,1,null,null],
  [8,null,null,9,null,7,null,null,3,null,null,null,1,4,6,null,null,null,null,null,9,null,null,null,1,null,null,2,3,null,6,null,5,null,1,7,null,6,null,null,null,null,null,4,null,9,1,null,4,null,8,null,3,6,null,null,3,null,null,null,6,null,null,null,null,null,5,3,9,null,null,null,4,null,null,7,null,2,null,null,1],
  [3,7,null,null,null,null,8,1,null,6,null,null,3,null,null,null,4,5,null,null,null,1,null,null,7,null,2,null,5,3,4,null,null,null,2,null,null,null,null,null,null,null,null,null,null,null,2,null,null,null,7,4,8,null,5,null,7,null,null,3,null,null,null,9,1,null,null,null,6,null,null,3,null,3,4,null,null,null,null,6,8],
  [1,null,null,2,null,5,null,null,6,null,null,null,8,7,4,null,null,null,null,null,5,null,9,null,8,null,null,8,2,null,null,null,null,null,3,5,null,5,3,null,null,null,6,7,null,4,7,null,null,null,null,null,9,8,null,null,8,null,1,null,5,null,null,null,null,null,7,8,9,null,null,null,3,null,null,5,null,2,null,null,9],
  [null,3,6,null,null,1,null,null,null,null,null,null,3,null,2,4,null,null,null,8,null,null,7,null,5,null,null,null,null,9,null,null,4,null,null,2,1,null,null,null,null,null,null,null,6,2,null,null,7,null,null,3,null,null,null,null,1,null,8,null,null,3,null,null,null,4,5,null,3,null,null,null,null,null,null,1,null,null,7,2,null],
  [4,null,null,null,1,null,9,2,null,null,5,null,null,null,null,null,null,4,null,null,null,7,9,4,3,null,5,2,null,7,null,3,null,5,null,null,5,null,4,2,null,9,6,null,1,null,null,1,null,8,null,7,null,2,3,null,2,6,4,7,null,null,null,6,null,null,null,null,null,null,1,null,null,9,5,null,2,null,null,null,3],
  [null,null,4,1,2,5,8,null,null,null,null,null,null,4,null,null,null,null,8,null,null,7,null,9,null,null,5,4,null,5,null,null,null,1,null,3,9,6,null,null,null,null,null,4,7,2,null,1,null,null,null,5,null,8,6,null,null,9,null,4,null,null,1,null,null,null,null,7,null,null,null,null,null,null,2,3,1,6,7,null,null],
  [8,9,null,null,null,4,null,null,7,3,1,5,null,null,null,null,2,null,null,6,7,null,null,null,8,null,null,null,null,null,6,4,null,null,null,1,null,null,null,3,7,2,null,null,null,7,null,null,null,8,1,null,null,null,null,null,4,null,null,null,2,8,null,null,8,null,null,null,null,1,7,5,1,null,null,5,null,null,null,9,6],
  [null,null,8,null,null,null,6,null,null,null,7,null,null,1,null,null,5,null,5,null,null,3,null,6,null,null,4,null,null,2,6,null,3,4,null,null,null,3,null,null,null,null,null,6,null,null,null,6,5,null,7,9,null,null,9,null,null,7,null,2,null,null,8,null,6,null,null,9,null,null,3,null,null,null,7,null,null,null,2,null,null],
  [null,5,null,8,null,1,null,9,null,7,3,null,null,null,null,null,5,4,8,null,null,null,3,null,null,null,1,null,null,8,3,null,2,1,null,null,null,null,null,null,null,null,null,null,null,null,null,6,7,null,4,5,null,null,1,null,null,null,5,null,null,null,9,3,8,null,null,null,null,null,1,2,null,4,null,6,null,8,null,3,null],
  [null,null,3,null,5,null,null,1,null,null,null,null,null,null,null,9,4,null,5,null,null,null,null,2,null,null,8,null,2,null,3,4,1,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,8,9,7,null,6,null,3,null,null,5,null,null,null,null,6,null,6,1,null,null,null,null,null,null,null,7,null,null,2,null,4,null,null],
];
