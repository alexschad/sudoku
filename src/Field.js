import React, { useState, useEffect, useMemo } from 'react'
import { rows, cols, squares, sudokus } from './data';

const top = [0,1,2,3,4,5,6,7,8,
  27,28,29,30,31,32,33,34,35,
  54,55,56,57,58,59,60,61,62];
const bottom = [72,73,74,75,76,77,78,79,80];
const left = [0,3,6,9,12,15,18,
  21,24,27,30,33,36,39,42,45,
  48,51,54,57,60,63,66,69,72,75,78];
const right = [8,17,26,35,44,53,62,71,80];

const getFieldClass = (index, isFixed, over, highlight, errHighlight, errNumber) => {
  const classes = ['field'];
  if (over) classes.push('over');
  if (highlight) classes.push('highlight');
  if (errHighlight) classes.push('errHighlight');
  if (errNumber) classes.push('errNumber');

  if (top.includes(index)) {
    if (left.includes(index)) {
      classes.push('topleft')
    } else if (right.includes(index)) {
      classes.push('topright');
    } else {
      classes.push('top');
    }
  }

  if (bottom.includes(index)) {
    if (left.includes(index)) {
      classes.push('bottomleft')
    } else if (right.includes(index)) {
      classes.push('bottomright');
    } else {
      classes.push('bottom');
    }
  }

  if (left.includes(index) &&
    !top.includes(index) &&
    !bottom.includes(index)) classes.push('left');

  if (right.includes(index) &&
    !top.includes(index) &&
    !bottom.includes(index)) classes.push('right');
  if (isFixed) classes.push('fixed');
  return classes.join(' ');
};

export default function Field({index, fields, sudokuIndex, showSelector, showHints, mouseEnter, mouseLeave, over, highlight=false, errHighlight=false, errNumber=false}) {
  const [usedVals, setUsedVals] = useState([]);
  const [isFixed, setIsFixed] = useState([]);

  useEffect(() => {
    setIsFixed(sudokuIndex !== null ? sudokus[sudokuIndex][index] !== null : false);
  },[sudokuIndex, index]);

  const cssClass = useMemo(() => getFieldClass(index, isFixed, over, highlight, errHighlight, errNumber),
  [index, isFixed, over, highlight, errHighlight, errNumber]);  

  useEffect(() => {
    const row = rows.find(r => r.includes(index));
    const col = cols.find(c => c.includes(index));
    const square = squares.find(s => s.includes(index));
    const usedIndices = (row && col && square) ? row.concat(col).concat(square) : [];
    let usedVals = usedIndices.map(ui => fields[ui]).filter(i => i !== null);
    usedVals = new Set(usedVals);  
    usedVals = [...usedVals];
    setUsedVals(usedVals);
  },[index, fields]);

  return (
    <div className={cssClass}
      onMouseEnter={() => mouseEnter(index)}
      onMouseLeave={() => mouseLeave()}
      onClick={isFixed ? () => {} : (e) => showSelector(e, index)}>
      {fields[index]}
      {!isFixed && showHints && fields[index] === null ? <div className="hint">{9 - usedVals.length}</div> : ''}
    </div>
  )
}


