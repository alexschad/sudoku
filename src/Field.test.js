import React from 'react';
import Field from './Field';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { sudokus } from './data';

test('it renders top left', () => {
  const component = renderer.create(
    <Field
    key={0}
    index={0}
    fields={sudokus[0]}
    sudokuIndex={0}
    showSelector={() => {}}
    showHints={false}
    mouseEnter={() => {}}
    mouseLeave={() => {}}
    highlight={true}
    errHighlight={false}
    errNumber={false}
    over={false}
  />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('it renders top center', () => {
  const component = renderer.create(
    <Field
    key={0}
    index={3}
    fields={sudokus[0]}
    sudokuIndex={0}
    showSelector={() => {}}
    showHints={false}
    mouseEnter={() => {}}
    mouseLeave={() => {}}
    highlight={true}
    errHighlight={false}
    errNumber={false}
    over={false}
  />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('it renders center left', () => {
  const component = renderer.create(
    <Field
    key={0}
    index={18}
    fields={sudokus[0]}
    sudokuIndex={0}
    showSelector={() => {}}
    showHints={false}
    mouseEnter={() => {}}
    mouseLeave={() => {}}
    highlight={true}
    errHighlight={false}
    errNumber={false}
    over={false}
  />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('it renders center right', () => {
  const component = renderer.create(
    <Field
    key={0}
    index={17}
    fields={sudokus[0]}
    sudokuIndex={0}
    showSelector={() => {}}
    showHints={false}
    mouseEnter={() => {}}
    mouseLeave={() => {}}
    highlight={true}
    errHighlight={false}
    errNumber={false}
    over={false}
  />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('it renders top right', () => {
  const component = renderer.create(
    <Field
    key={0}
    index={8}
    fields={sudokus[0]}
    sudokuIndex={0}
    showSelector={() => {}}
    showHints={false}
    mouseEnter={() => {}}
    mouseLeave={() => {}}
    highlight={true}
    errHighlight={false}
    errNumber={false}
    over={false}
  />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('it renders top center', () => {
  const component = renderer.create(
    <Field
    key={0}
    index={1}
    fields={sudokus[0]}
    sudokuIndex={0}
    showSelector={() => {}}
    showHints={false}
    mouseEnter={() => {}}
    mouseLeave={() => {}}
    highlight={true}
    errHighlight={false}
    errNumber={false}
    over={false}
  />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('it renders bottom left', () => {
  const component = renderer.create(
    <Field
    key={0}
    index={72}
    fields={sudokus[0]}
    sudokuIndex={0}
    showSelector={() => {}}
    showHints={false}
    mouseEnter={() => {}}
    mouseLeave={() => {}}
    highlight={true}
    errHighlight={false}
    errNumber={false}
    over={false}
  />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('it renders bottom right', () => {
  const component = renderer.create(
    <Field
    key={0}
    index={80}
    fields={sudokus[0]}
    sudokuIndex={0}
    showSelector={() => {}}
    showHints={false}
    mouseEnter={() => {}}
    mouseLeave={() => {}}
    highlight={true}
    errHighlight={false}
    errNumber={false}
    over={false}
  />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('it renders bottom center', () => {
  const component = renderer.create(
    <Field
    key={0}
    index={73}
    fields={sudokus[0]}
    sudokuIndex={0}
    showSelector={() => {}}
    showHints={false}
    mouseEnter={() => {}}
    mouseLeave={() => {}}
    highlight={true}
    errHighlight={false}
    errNumber={false}
    over={false}
  />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('test mouse events', () => {
  const component = renderer.create(
    <Field
    key={0}
    index={73}
    fields={sudokus[0]}
    sudokuIndex={0}
    showSelector={() => {}}
    showHints={false}
    mouseEnter={() => {}}
    mouseLeave={() => {}}
    highlight={true}
    errHighlight={false}
    errNumber={false}
    over={false}
  />,
  );
  let tree = component.toJSON();

  const mockEnterCallBack = jest.fn();
  const mockLeaveCallBack = jest.fn();
  const mockShowSelector = jest.fn();
  
  // mouse enter
  const field = shallow((<Field
    key={0}
    index={73}
    fields={sudokus[0]}
    sudokuIndex={0}
    showSelector={mockShowSelector}
    showHints={false}
    mouseEnter={mockEnterCallBack}
    mouseLeave={mockLeaveCallBack}
    highlight={true}
    errHighlight={false}
    errNumber={false}
    over={false}
  />));
  field.simulate('MouseEnter');
  expect(mockEnterCallBack.mock.calls.length).toEqual(1);

  field.simulate('mouseleave');
  expect(mockLeaveCallBack.mock.calls.length).toEqual(1);

  field.simulate('click');
  expect(mockShowSelector.mock.calls.length).toEqual(1);

});
