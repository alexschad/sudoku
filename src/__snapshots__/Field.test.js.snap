// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders bottom center 1`] = `
<div
  className="field highlight bottom"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  9
  
</div>
`;

exports[`it renders bottom left 1`] = `
<div
  className="field highlight bottomleft"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  
</div>
`;

exports[`it renders bottom right 1`] = `
<div
  className="field highlight bottomright"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  
</div>
`;

exports[`it renders center left 1`] = `
<div
  className="field highlight left"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  7
  
</div>
`;

exports[`it renders center right 1`] = `
<div
  className="field highlight right fixed"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  7
  
</div>
`;

exports[`it renders top center 1`] = `
<div
  className="field highlight topleft"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  
</div>
`;

exports[`it renders top center 2`] = `
<div
  className="field highlight top"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  
</div>
`;

exports[`it renders top left 1`] = `
<div
  className="field highlight topleft"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  
</div>
`;

exports[`it renders top right 1`] = `
<div
  className="field highlight topright"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  
</div>
`;
